### ISSUE TITLE
Questions/suggestions

### ISSUE THREAD

**Issue by georgiosd** [AngleSharp/AngleSharp#561]
Hello again,

I have several questions and some suggestions. If we can agree on the route to take, I will send you PRs for at least some of them :)

1. Missing feature: handling of meta-refresh
2. Suggestion: Implement a callback with cancellation support for redirects and reloads (whether JS or meta). Also a configuration flag to allow redirects/reloads
3. Suggestion: Provide a backwards-compatible way to inject a request filter interface .Filter is not always available when I create the context and I want to be able to change it throughout. Which would mean reinstating public methods to replace services in `IConfiguration`
4. Issue: The `MemoryCookieProvider` needs a lot of work+tests. I've had several problems with cookies that have `UTC` instead of `GMT` and cases where the cookies are not returned by the `CookieContainer` because they were added for a path by `AngleSharp` even though `path=/` for the cookie itself. It also chokes with websites that redirect from `http://foo.com` to `http://www.foo.com` later in a workflow.
5. Improvement: Replace network tests with tests that mock `IRequester`
6. Issue: loading `twitter.com` never returns
7. Issue: Some forms end up with submitting duplicate fields, somehow. It happens with Chrome also, but I somehow feel AngleSharp should do better :)
8. Question: Do responses/IDocuments get disposed? When, how? When I was parsing myself i would dispose the doc as soon as I was done.
9. Missing: some overloads for e.g. `OpenAsync(DocumentRequest, default(CancellationToken))`
10. Question: Does session history work? Array seems to be null/empty.
11. Nice-to-have: Request interceptor!
12. Question/Missing: How does one simulate an XHR request??

Hope I didn't overload you but I think that's everything, before I forget :)

**Comment by FlorianRappl** [comment]
I think most of these things are beyond this single library, but let's tackle the points in order:

1. There should be a generic meta provider, but that can be implemented on top with ease and should not be in the core. Could be part of AngleSharp.Browser or some other external lib.
2. Can be done already by exchanging the used loader. The loader is quite generic and thus should not be mixed with custom needs. Same approach as with 1.
3. See 2.
4. The `MemoryCookieProvider` is a simple implementation of a cookie provider. Most problems come actually from the .NET cookie container. In my opinion the solution is to create a much better one, however, I would not see this one in the core. The core should be as lightweight but complete as possible. See 1.
5. Actually, most tests mock this. Only a couple of integration tests use this and all of these cases are relevant. A (self-controlled) test server is used for some of these days. The actual improvement I see is moving all external integration tests to this self-controlled test server.
6. Separate issue. Could be anything from an HTTP(S) problem to CSS or scripting. Without knowledge, e.g., about the used configuration there is not much I can write...
7. Potentially due to the spec. First read the spec if you think something fishy is going on. If Chrome does it there is a good chance this is intentional... But we also had a bug in AngleSharp that was actually also a bug in Chrome (identified and reported that way).
8. You are responsible for disposing `IDocument`.
9. Sounds reasonable.
10. This is the core - the context implementation is only minimal and has no session history. See 1.
11. Not sure what is meant with interceptor. There is a filter which you mentioned and there are events for the two phases. Other than that you can always provide an `IRequester` implementation that uses an original implementation and thus is an interceptor.
12. The original XHR API is implemented in the Scripting library. Other than that the virtual response API may be what you are looking for.

Regarding an advanced cookie provider the existing `AngleSharp.Io` library could also be seen as an ideal fit. There is also a good chance that it may be a good candidate for some pre-made helpers / constructs to have something like a request interceptor or other `IRequester` implementations.

**Comment by georgiosd** [comment]
Ok, so I guess I'll keep off the PRs and implement most of this privately :)

Re:
5. To avoid actually sending remote request (I did see your azurewebsites site), I use Titanium Proxy to fake a response, as saved from a Fiddler SAZ file. Perhaps of some value to you.
6. From what I can tell, it has something to do with XHRs generated by twitter JS that refresh the feed
7. You mean the spec of the form? Pretty sure it's got those fields only once.
10. How would you implement the session history outside of the context though? And that would mean keeping all the IDocuments in memory until the context is disposed right?
11. I meant an interceptor service that will receive calls when any request is completed (i.e. regardless of the IRequester that will serve it)
12. Not sure if we meant the same thing - what I mean is, how to provide XHR semantics - i.e. the request is composed by "hand" and the response is given more "raw" (would most likely be JSON). Also redirects are not followed.

**Comment by FlorianRappl** [comment]
5. It's certainly a good solution, but for actually faking responses there is already a dedicated requester in the test code. It will save the response (if not available) and serve the saved response.
7. The W3C specification of sending form fields. Sending fields multiple times is actually valid (as you should know) so this cannot be the issue you are referring to. If you think part of the spec has been implemented incorrectly then please present a MWE to reproduce your issue.
10. There would be two parts: One is the session history and the other one is an improved browsing context. However, the latter is not needed. Yes, it would potentially keep the documents in memory, however, not the full tree. This is all defined by the W3C specification again and the unload event is still making sure web app creators can do a proper cleanup when the active document changes.
11. I think the loader solves this as the loader is calling the requester.
12. The `IRequester` is that. In general it is not supposed to follow redirects. The standard loader thus recursively calls the requester until we either have exceeded the max. redirects or end up at a non-redirect response.

**Comment by georgiosd** [comment]
Gotcha. I'll check the form again and let you know.

By the way, using the proxy it's a little different than mocking the IRequester - it's effectively an integration test because the WebRequest actually happens. 

Of course another way would be to self-host a Web API with all the responses you need to run the tests.

**Comment by FlorianRappl** [comment]
> By the way, using the proxy it's a little different than mocking the IRequester - it's effectively an integration test because the WebRequest actually happens.

I agree but my point was a different one: For simple requests where we only care about the response the special requester that loads the response from a saved instance is sufficient (no need to include a third party lib for that).

**Comment by georgiosd** [comment]
Ok.

On the issue of the project.json creating problems, have you found a solution? 
It seems to be resolved by this commit: https://github.com/georgiosd/AngleSharp/commit/b5af9378f94f3a3738673079e39cfcbf9a5529a6

But I'm not sure if that's kosher with you.

**Comment by georgiosd** [comment]
Another thing I just found:

It would be good, at least in `DEBUG`, to somehow report the exceptions happening in the background tasks (see `TaskEventLoop`). How about providing a delegate for it?

**Comment by FlorianRappl** [comment]
I think the proper way to solve the building issue is to migrate back to csproj using the .NET Core migration tooling. I could not look into this...

Unfortunately, I can't accept that one commit as it just removes the rest and goes to a portable library (which is what you already get from the standard csproj originally delivered here). The multi-target build is quite important (thus supporting SL5, .NET 4, ...).

I like the idea of a delegate, however, I think it is something that should not be generic. It should be specific to an event loop. So for the very simple default task event loop we can introduce either an event or a property accepting a delegate to handle this.

**Comment by georgiosd** [comment]
All that .NET standard stuff baffles me so I can't help further :)

Re the delegate, I was thinking something like `TaskScheduler.UnobservedException` callback. I mean, whether you use `Tasks` or not, the event loop is usually running in a different thread and exceptions there have to somehow make it out of there.

So in this case I would suggest something like `BrowsingContext.EventLoopException`

**Comment by FlorianRappl** [comment]
Yeah, the only thing is that the browsing context has no knowledge about the event loop implementation. And the ideal event loop would use a pure reactor pattern, i.e., all the running code is in a single thread, whereas the event emitters (e.g., the HTTP requester) may be on *any* other thread.

That all being written an implementation could of course access the field you just described, however, I would like to bring the exception handling delegate to the origin (i.e., in this case the task-based event loop) to not overlap responsibilities. A user can still get this by obtaining a reference to the specific service and setting the delegate. Granted, that requires more knowledge than just using the `BrowsingContext`, but keep in mind that a user usually only interacts with the `IBrowsingContext` and would need to cast anyway (unless you also bring the property to the interface, but then we have a serious issue anyway by requiring to add this property in whatever implementation).

Long story short: I still have the opinion that this delegate would be best exposed at the event loop implementation itself.

**Comment by georgiosd** [comment]
Okay cool. Why don't you pick a name for the event handler and I'll add it to `TaskEventLoop` before you merge the other changes.

One more thing I just remembered: is there any reason why you are holding an `IEnumerable<object>` reference in `Configuration`? Since it's immutable, it would be safer to hold an `object[]` - if someone passes an actual `IEnumerable<object>` there's the danger that it will be enumerated several times. And since the configuration object is not modified a lot, the overhead of creating arrays for every change should be negligible.

I also saw that you removed access to the `Configuration` object in `BrowsingContext`. I agree it's cleaner but now I don't have the capability to replace services throughout the lifetime of the context. Any smart solutions you can think of? :)

**Comment by georgiosd** [comment]
Closing this and moving conversation directly to the PR: https://github.com/AngleSharp/AngleSharp/pull/562
